### PR TITLE
PLAT-1331-3 Enable StoredFileFileSystemContentStore

### PR DESCRIPTION
**Pull request by alexander-schmidt**
- Stored Files can now be saved to either SMB shares or local file system directories.
  - An appropriate `IStoredFileContentStore` implementation is chosen according to the protocol part of the repository URL, as defined in `Core`.`StoredFileRepository`.
  - e.g. `StoredFileSmbContentStore` will be used for URLs like `smb://host/share`
  - e.g. `StoredFileFileSystemContentStore` will be used for URLs like `file:///dir/subdir`
- Established `StoredFileContentStores` as a central place to retrieve (SMB-based or file-system-based) `IStoredFileContentStore` instances.
  - This puts an end to seemingly-random calls to file store implementation constructors throughout the code base.
- `IStoredFileContentStore`: merged `isEnabled`, `isReady` and `isAvailable` to just one method, `isAvailable`.
  - All existing implementations of `isEnabled` were either trivial or nonsensically-arbitrary.
- `StoredFileContentInputStreamCreator` now checks whether a file store is available before trying to access it.

### CODE REVIEW

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStoreFactory.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+				Log.ferror("Cannot handle content store URLs that start with '%s'.", protocol);
+			}
+		} catch (MalformedURLException exception) {
+			Log.ferror("Invalid file repository URL: %s", StackTraceFormatting.getStackTraceAsString(exception));
+		}
+
+		return Optional.empty();
+	}
+
+	private static class SimpleUrl {
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStoreFactory.java:33]: `java.net.URL` throws for URLs that start with `smb`, so I built this.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStoreFactory.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		public SimpleUrl(String text) throws MalformedURLException {
+
+			var matcher = PATTERN.matcher(text);
+			matcher.find();
+			if (matcher.groupCount() != 2) {
+				throw new MalformedURLException("Illegal URL: '%s'".formatted(text));
+			}
+
+			this.protocol = matcher.group(1).toLowerCase();
+			this.path = Trim.trimPrefix(matcher.group(2), "//");
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStoreFactory.java:48]: Note that the following URLs are all valid, in principle:
```
file:/dir/subdir    - an absolute path (short form)
file://dir/subdir   - a relative path
file:///dir/subdir  - an absolute path
```

File: platform-dom/src/main/resources/com/softicar/platform/dom/dom-wiki-elements-style.css [hunk 3]
```diff
@@ -4,12 +4,16 @@ h6.DomWikiHeadline {
...
+	margin-top: 0;
+}
+
 .DomWikiMonospace {
 	display: inline;
 	color: #880000;
 	background-color: #f0f0f0;
 	border-radius: var(--BORDER_RADIUS);
-	box-sahdow: 0 0 0 1px #cccccc inset;
+	box-shadow: 0 0 0 1px #cccccc inset;
```
Comment (alexander-schmidt) [inline platform-dom/src/main/resources/com/softicar/platform/dom/dom-wiki-elements-style.css:16]: Fixed random typo / broken declaration

File: platform-dom/src/main/resources/com/softicar/platform/dom/dom-wiki-elements-style.css [hunk 4]
```diff
@@ -4,12 +4,16 @@ h6.DomWikiHeadline {
 	border-bottom: 1px solid var(--LABEL_BORDER_COLOR);
 }
 
+.DomWikiList {
+	margin-top: 0;
```
Comment (alexander-schmidt) [inline platform-dom/src/main/resources/com/softicar/platform/dom/dom-wiki-elements-style.css:8]: Before:
image

After:
image

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploader.java [hunk 5]
```diff
@@ -32,35 +33,43 @@ class StoredFileContentUploader {
...
 	public OutputStream createOutputStream() {
 
-		if (!store.isEnabled()) {
+		if (!store.isPresent()) {
 			return useChunks();
 		}
 
-		if (store.isReady()) {
-			long freeDiskSpace = store.getFreeDiskSpace();
+		if (store.get().isAvailable()) {
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploader.java:57]: Abundant use of `store.get()` here and below should probably be fixed / beautified.

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/cleanup/StoredFilesCleanupProgram.java [hunk 6]
```diff
@@ -4,33 +4,38 @@
...
 
-		StoredFileSmbContentStore store = new StoredFileSmbContentStore();
+		var removeAtThreshold = DayTime.now();
+		var storedFileDatabase = new StoredFileDatabase();
+		var primaryContentStore = StoredFileContentStores.getPrimaryContentStore();
 
-		new FileStoreTmpFolderCleaner(store).cleanAll();
-
-		DayTime removeAtThreshold = DayTime.now();
+		primaryContentStore.ifPresent(store -> new FileStoreTmpFolderCleaner(store).cleanAll());
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/cleanup/StoredFilesCleanupProgram.java:30]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentOutputStreamCreator.java [hunk 7]
```diff
@@ -6,38 +6,40 @@
...
 	private final IStoredFileDatabase database;
-	private IStoredFileContentStore store;
+	private Optional<IStoredFileContentStore> store;
 
 	public StoredFileContentOutputStreamCreator(AGStoredFile storedFile) {
 
 		this.storedFile = storedFile;
 		this.database = new StoredFileDatabase();
-		this.store = new StoredFileSmbContentStore();
+		this.store = StoredFileContentStores.getPreferredAvailableContentStore();
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentOutputStreamCreator.java:31]: We could, for sure, remove the fallback to secondary file stores here. However, the fallback to database-based file storage (which is not available as an `IStoredFileContentStore`) would have to remain in `StoredFileContentUploader`.
Ack?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploader.java [hunk 8]
```diff
@@ -32,35 +33,43 @@ class StoredFileContentUploader {
...
 	private final AGStoredFile storedFile;
+	private final Optional<IStoredFileContentStore> store;
 
-	public StoredFileContentUploader(IStoredFileDatabase database, IStoredFileContentStore store, AGStoredFile storedFile) {
+	public StoredFileContentUploader(IStoredFileDatabase database, AGStoredFile storedFile, IStoredFileContentStore store) {
+
+		this(database, storedFile, Optional.ofNullable(store));
+	}
+
+	public StoredFileContentUploader(IStoredFileDatabase database, AGStoredFile storedFile, Optional<IStoredFileContentStore> store) {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploader.java:44]: Added non-null requirements and removed `Optional`-based CTOR.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploader.java [hunk 9]
```diff
@@ -105,26 +114,25 @@ private void moveFile(byte[] hash) {
 
 	private void createFolders(String folderName) {
 
-		store.createDirectories(folderName);
+		store.get().createDirectories(folderName);
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploader.java:117]: Granted - yet, the Java API pretty much consistently uses `directory`. So consistently using the latter in our code as well should eliminate unnecessary synonyms.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/IStoredFileContentStore.java [hunk 10]
```diff
@@ -28,31 +28,13 @@ public interface IStoredFileContentStore {
...
-
-	/**
-	 * @deprecated use {@link #isReady()} instead
-	 */
-	@Deprecated
-	default boolean isAvailable() {
-
-		return isReady();
-	}
+	boolean isAvailable();
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/IStoredFileContentStore.java:37]: Opted for `isAccessible`.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStoreFactory.java (hunk 2 above)
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStoreFactory.java:48]: Added a Javadoc block above `SimpleUrl`.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStores.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+	 * to {@link IStoredFileContentStore#isAvailable()}), that content store
+	 * will be the first element in the returned {@link Collection}.
+	 * <p>
+	 * If no content store can be determined, an empty {@link Collection} is
+	 * returned.
+	 *
+	 * @return all currently available content stores
+	 * @see IStoredFileContentStore#isAvailable()
+	 */
+	public static Collection<IStoredFileContentStore> getAvailableContentStores() {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStores.java:34]: Indeed, `accessible` is better.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStores.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+	 * will be returned. Otherwise, an available non-primary content store may
+	 * be returned.
+	 * <p>
+	 * If no content store can be determined, {@link Optional#empty()} is
+	 * returned.
+	 *
+	 * @return the currently preferred and available content store
+	 * @see IStoredFileContentStore#isAvailable()
+	 */
+	public static Optional<IStoredFileContentStore> getPreferredAvailableContentStore() {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/store/StoredFileContentStores.java:56]: we don't. it was removed.

File: platform-core-module/src/test/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploaderTest.java [hunk 13]
```diff
@@ -129,41 +128,19 @@ public void movesAndRenamesAccordingToHash() {
...
 
 		upload();
 
 		Mockito.verify(database).createChunksOutputStream(ArgumentMatchers.same(storedFile));
 		DbAssertUtils.assertOne(AGLogMessage.TABLE);
 	}
 
-	@SuppressWarnings("resource")
-	@Test
-	public void usesDatabaseOutputStreamIfFileStoreDisabled() {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentUploaderTest.java:142]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentOutputStreamCreator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/file/stored/content/StoredFileContentOutputStreamCreator.java]: that is fine, IMO

**Pull request merged**